crates/sui-rest-api/src/transactions/resolve.rs

```diff
@@ -0,0 +1,564 @@
+
+    ProgrammableTransaction {
+        inputs,
+        commands: unresolved_ptb
+            .commands
+            .into_iter()
+            .map(Into::into)
+            .collect(),
+    }
+    .pipe(Ok)
+}
+
+fn resolve_arg(
+    reader: &StateReader,
+    called_packages: &HashMap<ObjectId, NormalizedPackage>,
+    commands: &[Command],
+    arg: UnresolvedInputArgument,
+    arg_idx: usize,
+) -> Result<CallArg> {
+    match arg {
+        UnresolvedInputArgument::Pure { value } => CallArg::Pure(value),
+        UnresolvedInputArgument::ImmutableOrOwned(obj_ref) => CallArg::Object(
+            ObjectArg::ImmOrOwnedObject(resolve_object_reference(reader, obj_ref)?),
+        ),
+        UnresolvedInputArgument::Shared {
+            object_id,
+            initial_shared_version: _,
+            mutable: _,
+        } => {
+            let id = object_id.into();
+            let object = reader
+                .inner()
+                .get_object(&id)?
+                .ok_or_else(|| ObjectNotFoundError::new(object_id))?;
+
+            let initial_shared_version = if let sui_types::object::Owner::Shared {
+                initial_shared_version,
+            } = object.owner()
+            {
+                *initial_shared_version
+            } else {
+                return Err(RestError::new(
+                    axum::http::StatusCode::BAD_REQUEST,
+                    format!("object {object_id} is not a shared object"),
+                ));
+            };
+
+            let mut mutable = false;
+
+            for (move_call, idx) in find_arg_uses(arg_idx, commands).filter_map(|(command, idx)| {
+                if let Command::MoveCall(move_call) = command {
+                    idx.map(|idx| (move_call, idx))
+                } else {
+                    None
+                }
+            }) {
+                let function = called_packages
+                    // Find the package
+                    .get(&move_call.package)
+                    // Find the module
+                    .and_then(|package| package.normalized_modules.get(move_call.module.as_str()))
+                    // Find the function
+                    .and_then(|module| module.functions.get(move_call.function.as_str()))
+                    .ok_or_else(|| {
+                        RestError::new(
+                            axum::http::StatusCode::BAD_REQUEST,
+                            "unable to find function",
+                        )
+                    })?;
+
+                let arg_type = function.parameters.get(idx).ok_or_else(|| {
+                    RestError::new(
+                        axum::http::StatusCode::BAD_REQUEST,
+                        "invalid input parameter",
+                    )
+                })?;
+
+                if matches!(
+                    arg_type,
+                    move_binary_format::normalized::Type::MutableReference(_)
+                ) {
+                    mutable = true;
```

Comment:
should be able to break out of the loop early after setting mutable to true

Reply:
Because of shared object deletion, you also need to look for functions that accept the shared object by value. If you have any of those, then it's also considered mutable.